feat(dashboard): enabling multi region setup from single dashboard and clerk

merrcury opened the pull request:
This pull request introduces support for region-aware API and WebSocket endpoints throughout the dashboard application, including the addition of a region selector to the authentication and navigation flows. It refactors several components to use a new `apiHostnameManager` utility for dynamic hostname resolution, ensures telemetry captures region metadata during organization creation, and updates the deployment workflow to support a new `staging-apse1` environment. Minor code style improvements (such as consistent use of double quotes) are also included.

**Region-aware API and WebSocket endpoints:**

* Refactored multiple components (`api.client.ts`, `inbox-preview-content.tsx`, `inbox-button.tsx`, `configuration-group.tsx`, `simple-prompt-getter.ts`) to use `apiHostnameManager` for dynamic selection of API and WebSocket hostnames, enabling region switching and improved reliability during region changes. [[1]](diffhunk://#diff-fad440368ea23c6b58e7f3b8d8c244d6a66a26ca8fc2acc804ea7c248139f880L1-R3) [[2]](diffhunk://#diff-fad440368ea23c6b58e7f3b8d8c244d6a66a26ca8fc2acc804ea7c248139f880R41-R45) [[3]](diffhunk://#diff-fad440368ea23c6b58e7f3b8d8c244d6a66a26ca8fc2acc804ea7c248139f880L65-R69) [[4]](diffhunk://#diff-98f1b0ca39b055dd7eafe223b865bee6ae2f25c53254829bfa745136b151dbdcR1-L3) [[5]](diffhunk://#diff-98f1b0ca39b055dd7eafe223b865bee6ae2f25c53254829bfa745136b151dbdcL35-R36) [[6]](diffhunk://#diff-b1745a3795d78b3df0f9fd630e87e2e2688c8e573147de73cb229e2f96aae412L1-R9) [[7]](diffhunk://#diff-b1745a3795d78b3df0f9fd630e87e2e2688c8e573147de73cb229e2f96aae412L118-R127) [[8]](diffhunk://#diff-5b7ed7880590b045a4bdbfcd29adae6ac028a3871164f34ad070eec50b69c04bR1-R7) [[9]](diffhunk://#diff-5b7ed7880590b045a4bdbfcd29adae6ac028a3871164f34ad070eec50b69c04bL13-R27) [[10]](diffhunk://#diff-94f501d2219526355cd1e7049ed400a1a7da4f6ceb5929b6b51c28f92de7cf11L1-R2) [[11]](diffhunk://#diff-94f501d2219526355cd1e7049ed400a1a7da4f6ceb5929b6b51c28f92de7cf11L43-R44)

**Region selector integration:**

* Added a `RegionSelector` component to the organization creation flow (`create-organization.tsx`), header navigation (`header-navigation.tsx`), and telemetry tracking, allowing users to select their region during onboarding and navigation. [[1]](diffhunk://#diff-d5fc6207f2a7bca539d051b6567a4cb9c4a238f46573e70ef8303258f6ed2e37R1-R3) [[2]](diffhunk://#diff-d5fc6207f2a7bca539d051b6567a4cb9c4a238f46573e70ef8303258f6ed2e37L63-R98) [[3]](diffhunk://#diff-d5fc6207f2a7bca539d051b6567a4cb9c4a238f46573e70ef8303258f6ed2e37R151) [[4]](diffhunk://#diff-d5fc6207f2a7bca539d051b6567a4cb9c4a238f46573e70ef8303258f6ed2e37R161-R170) [[5]](diffhunk://#diff-97ba9f7335ac4672b0818defcd1c8fd7728a5d07cec1b484340751fc4c167e56L1-R9) [[6]](diffhunk://#diff-97ba9f7335ac4672b0818defcd1c8fd7728a5d07cec1b484340751fc4c167e56L57-R61)

**Deployment workflow enhancements:**

* Updated `.github/workflows/deploy.yml` to add support for the `staging-apse1` environment, including input options and environment matrix logic, and standardized YAML quoting for improved consistency. [[1]](diffhunk://#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3L22-R49) [[2]](diffhunk://#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3R88-R90) [[3]](diffhunk://#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3L195-R200) [[4]](diffhunk://#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3L205-R209) [[5]](diffhunk://#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3L306-R310) [[6]](diffhunk://#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3L329-R335) [[7]](diffhunk://#diff-28802fbf11c83a2eee09623fb192785e7ca92a3f40602a517c011b947a1822d3L353)

**Miscellaneous improvements:**

* Updated Dockerfile entrypoint for the API service to use the `SECRET_NAME` environment variable for improved flexibility.
* Synced subproject reference in `.source` file.

These changes collectively improve the dashboard's support for multi-region deployments and user experience around region selection and API reliability.

merrcury committed:
feat(region): implement region management with dynamic API and WebSocket hostnames

- Introduced a new `RegionProvider` to manage region selection and API hostname updates.
- Added `RegionSelector` component for users to select their region.
- Integrated `apiHostnameManager` to handle API and WebSocket hostnames based on selected region.
- Updated various components and pages to utilize the new region context and hostname management.
- Implemented region detection from organization metadata for improved user experience.

merrcury committed:
chore: update subproject commit reference in .source

merrcury committed:
chore: update ENTRYPOINT in Dockerfiles to use dynamic SECRET_NAME variable

merrcury committed:
chore: update deploy.yml to enhance environment options and standardize descriptions

merrcury committed:
feat(region): enhance region detection and management

- Added URL-based region detection to replace localStorage approach for improved reliability.
- Introduced new constants for dashboard URLs specific to regions.
- Updated `RegionProvider` to utilize URL detection for initial region selection and organization switching.
- Enhanced `findOrganizationForRegion` and `getDashboardUrlForRegion` functions to support new detection logic.
- Improved logging for better debugging and user feedback during region changes.

merrcury committed:
refactor(region): streamline region redirection logic in RegionProvider

- Updated region handling to redirect users to the correct dashboard URL if the URL region does not match the organization's region.
- Improved logging for clarity on region mismatches and redirection actions.
- Simplified the logic for updating the selected region state when the URL and organization match.

merrcury committed:
refactor(auth, region): remove console logs and streamline organization creation flow

- Eliminated unnecessary console logs in `create-organization.tsx` and `region-context.tsx` for cleaner code and improved performance.
- Updated organization creation flow to focus on telemetry tracking without logging sensitive information.
- Enhanced region handling logic to ensure consistent URL redirection during organization creation and selection.

merrcury committed:
refactor(region): enhance logging and refine organization creation flow

- Improved debug logging in `RegionProvider` to provide better insights during region switching and organization creation.
- Updated logic to prevent redirection during organization creation when selecting existing organizations, ensuring a smoother user experience.
- Clarified comments and streamlined the handling of region mismatches for better maintainability.

merrcury committed:
refactor(config): simplify API and WebSocket hostname assignments

- Removed fallback logic for `API_HOSTNAME_SG` and `WEBSOCKET_HOSTNAME_SG`, directly assigning values from environment variables.
- Streamlined configuration for better clarity and maintainability.

merrcury committed:
refactor(config): simplify DASHBOARD_URL_SG assignment

- Removed fallback logic for `DASHBOARD_URL_SG`, directly assigning the value from environment variables for improved clarity and maintainability.

merrcury committed:
feat(region): implement multi-region configuration and dynamic loading

- Introduced a new `region-config.ts` file to manage multi-region setup, allowing dynamic loading of region-specific configurations from environment variables.
- Updated `RegionSelector` to dynamically generate region options based on the configured regions.
- Refactored region-related utility functions to utilize the new configuration system, enhancing maintainability and flexibility.
- Improved region detection logic to support multiple regions and fallback mechanisms.

merrcury committed:
refactor(region): enhance region configuration logic and defaults

- Updated region parsing to dynamically determine the base region from the VITE_REGIONS environment variable.
- Simplified the assignment of environment variables for dashboard URLs, API hostnames, and region flags based on whether the region is the base region or not.
- Improved comments for clarity on the configuration logic and fallback mechanisms for non-base regions.
- Enhanced logging to provide better insights when essential configuration is missing.

merrcury committed:
feat(env): add multi-region configuration to .example.env

- Introduced environment variables for multi-region support, allowing configuration for both US and Singapore regions.
- Added detailed comments to guide users on how to extend the configuration for additional regions.
- Defined base and additional region-specific URLs, API hostnames, and AWS regions for improved deployment flexibility.

merrcury committed:
refactor(region): remove outdated comments on region detection

- Eliminated comments regarding the localStorage-based region initialization in `region-context.tsx` and `region-utils.ts`, as the logic has been updated to rely solely on URL-based detection for improved reliability and clarity.

merrcury committed:
refactor(region): clean up comments and streamline region handling

- Removed outdated comments in `region-config.ts`, `region-context.tsx`, and `region-utils.ts` to enhance clarity and maintainability.
- Simplified comments related to environment variable usage and region mapping for better understanding of the configuration logic.
- Improved overall readability of the region handling code by eliminating unnecessary explanations.

scopsy commented on apps/dashboard/src/api/api.client.ts:
```diff
@@ -39,6 +38,11 @@ const request = async <T>(
 
   try {
+    // Prevent API calls during region switching to avoid failed requests
+    if (apiHostnameManager.isCurrentlyRegionSwitching()) {
+      console.log('Blocking API call during region switching');
+    }
```
Is this logic missing here?

merrcury replied:
Thanks for pointing this out. This issue came up when I first built everything on top of localStorage, which turned out to be unreliable. During the dashboard URL refactor, I switched it to use the console instead. Since the system now works properly with different dashboard URLs, this logic is no longer needed. I’ve removed both this code and the locks I had added to support it.

scopsy commented on apps/dashboard/src/components/auth/create-organization.tsx:
```diff
@@ -60,7 +61,41 @@ function FormContainer({ children }: FormContainerProps) {
+  return (
+    <div className="relative">
+      {/* Region selector - only visible on Page 2 (form page), aligned with form content */}
+      {showRegionSelector && (
+        <div className="absolute -top-14 left-4 z-20">
+          <RegionSelector />
+        </div>
```
Will this be visible to all users? Can we control this to specific users?

merrcury replied:
Its controlled via a LD flag in Selector component. I tested this with LaunchDarkly. Do you want me to bring it outside?
https://github.com/novuhq/novu/blob/f1d5b038117b804fb026e8d2ed9e35223fff0903/apps/dashboard/src/context/region/region-selector.tsx#L18-L19

scopsy commented on apps/dashboard/src/components/auth/create-organization.tsx:
```diff
@@ -122,9 +158,16 @@ export default function OrganizationCreate() {
+        region: selectedRegion,
       });
+
+      // Set the region metadata for the newly created organization
+      const regionMetadata = selectedRegion === 'singapore' ? 'ap-southeast-1' : 'us-east-1';
```
What is happening with this regionMetadata? Seems like it's defined but not used. Also, any specific reason for the hardcoding? Can we make it more dynamic?>

merrcury replied:
Thanks for pointing this out. This was originally used when the local code only supported two regions. Later, I expanded it to be dynamic, and at that time, we also needed metadata because we relied on localStorage to determine the API hostname for processing callbacks. Now, the hostname is decided directly from the URL, which is more reliable. So this code is no longer needed, and I’ve removed it.

scopsy commented on apps/dashboard/src/components/header-navigation/header-navigation.tsx:
```diff
@@ -54,8 +55,10 @@ export const HeaderNavigation = (props: HeaderNavigationProps) => {
+        <div className="flex items-center gap-2">
           <InboxButton />
+          <div className="h-4 w-px bg-neutral-200" />
+          <RegionSelector />
```
Can we control who sees it? Would prefer not to open this to everyone.

merrcury replied:
Same Its control via a LD flag

merrcury committed:
fix(region): improve API call handling during region switching

- Updated the API client to wait for region switching to complete before making API calls, enhancing reliability during transitions.
- Removed outdated comments in the organization creation flow to streamline the code and focus on telemetry tracking.
- Refined region switching logic in the RegionProvider to ensure accurate state management during region changes.

merrcury committed:
refactor(region): remove region switching logic from API client and context

- Eliminated the region switching checks from the API client, streamlining API call handling.
- Removed redundant region switching state management from the RegionProvider to enhance clarity and maintainability.
- Cleaned up the HostnameManager by removing the region switching state, simplifying the overall logic.

merrcury committed:
refactor(identity): improve user and organization identification logic

- Renamed identification flags for clarity and separated user and organization identification states.
- Enhanced user identification logic to ensure proper tracking with LaunchDarkly.
- Streamlined organization identification process, ensuring accurate monitoring and telemetry tracking.
- Cleaned up unused variables and improved overall readability of the IdentityProvider component.

merrcury committed:
refactor(region): enhance region name retrieval in organization creation modal

- Added a utility function to dynamically retrieve the region name based on the target region in the organization creation modal.
- Updated the modal description to use the new region name retrieval logic, improving clarity and maintainability of the code.

merrcury committed:
refactor(region): standardize default region handling in context and API keys

- Updated the organization creation modal to use a default region constant instead of hardcoded values, enhancing maintainability.
- Modified the API keys page to dynamically retrieve the region name based on the selected region configuration, improving clarity and flexibility in region handling.

merrcury committed:
update(subproject): update subproject commit reference to latest version

merrcury committed:
Merge branch 'next' of github.com:novuhq/novu into feat/singleClerk

merrcury committed:
update(subproject): change subproject commit reference to 15113238a9dc

merrcury committed:
update(subproject): change subproject commit reference to 6c74fe64b80f

merrcury committed:
chore(workflow): add NX_CLOUD_ACCESS_TOKEN environment variable to deploy workflow

merrcury committed:
refactor(docker): optimize Dockerfile and update build scripts for improved package handling

merrcury committed:
feat(workflow): add environment filtering to deployment logic in GitHub Actions

merrcury committed:
fix(workflow): update production environment names and conditions in deployment logic
